applications/FreeSurfaceApplication/custom_utilities/edgebased_levelset_auxiliary_utils.h

```diff
@@ -0,0 +1,458 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors: Riccardo Rossi
+//                Joaquín Irazábal
+//
+//
+
+#ifndef _EDGEBASED_LEVELSET_AUXILIARY_UTILS_H
+#define	_EDGEBASED_LEVELSET_AUXILIARY_UTILS_H
+/* System includes */
+
+
+/* External includes */
+
+
+/* Project includes */
+//#include "utilities/math_utils.h"
+#include "utilities/geometry_utilities.h"
+#include "utilities/math_utils.h"
+#include "utilities/body_distance_calculation_utils.h"
+
+
+namespace Kratos
+{
+
+/**@name Kratos Globals */
+/*@{ */
+
+
+/*@} */
+/**@name Type Definitions */
+/*@{ */
+
+
+/*@} */
+
+
+/**@name  Enum's */
+/*@{ */
+
+
+/*@} */
+/**@name  Functions */
+/*@{ */
+
+
+
+/*@} */
+/**@name Kratos Classes */
+/*@{ */
+
+/** Short class definition.
+Detail class definition.
+
+*/
+template< unsigned int TDim>
+class EdgebasedLevelsetAuxiliaryUtils
+{
+public:
+    /**@name Type Definitions */
+    /*@{ */
+    typedef ModelPart::NodesContainerType NodesArrayType;
+    typedef ModelPart::ElementsContainerType ElementsArrayType;
+    /*@} */
+    /**@name Life Cycle
+    */
+    /*@{ */
+
+    /** Constructor.
+    */
+
+    /** Destructor.
+    */
+
+    /*@} */
+    /**@name Operators
+    */
+    /*@{ */
+
+
+    /*@} */
+    /**@name Operations */
+    /*@{ */
+
+    void CalculateDistances(
+        ModelPart& r_model_part,
+        Variable<double>& rDistanceVar,
+        const double max_distance
+    )
+    {
+        if(TDim == 2)
+            CalculateDistances2D(r_model_part, rDistanceVar, max_distance);
+        else if (TDim == 3)
+            CalculateDistances3D(r_model_part, rDistanceVar, max_distance);
+
+     //   r_model_part.GetCommunicator().SynchronizeCurrentDataToMin(rDistanceVar);
+    }
+
+    //***********************************************************************
+    //***********************************************************************
+    void CalculateDistances2D(
+        ModelPart& r_model_part,
+        Variable<double>& rDistanceVar,
+        const double max_distance
+    )
+    {
+        KRATOS_TRY
+
+        double large_distance = 1e6;
+        double tol = 1.0/large_distance;
+
+        //copy rDistance var to GetValue database
+        for(ModelPart::NodesContainerType::iterator it =  r_model_part.NodesBegin(); it !=r_model_part.NodesEnd(); it++)
+        {
+            it->GetValue(rDistanceVar) = it->FastGetSolutionStepValue(rDistanceVar);
+            it->FastGetSolutionStepValue(rDistanceVar) = large_distance;
+            it->GetValue(IS_VISITED) = 0;
```

Comment:
For the sake of memory consumption, you can consider using the flag `VISITED` in here.